Add cuSPARSE descriptors and update CSR addmm

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62180
* #61858
* #61857
* #65606
* #64391
* #63948
* #63511
* #63510
* #63509
* #61536
* #61408
* #61407
* __->__ #60838

Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

IvanYashchuk committed:
Add cuSPARSE descriptors

Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors"


Rewrite addmm_out_sparse_csr_dense_cuda implementation

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk commented on torch/utils/hipify/cuda_to_hip_mappings.py:
```diff
@@ -554,6 +554,7 @@
         ),
         ("device_functions.h", ("hip/device_functions.h", CONV_INCLUDE, API_RUNTIME)),
         ("driver_types.h", ("hip/driver_types.h", CONV_INCLUDE, API_RUNTIME)),
+        ("library_types.h", ("hip/library_types.h", CONV_INCLUDE, API_RUNTIME)),
```
Changes from https://github.com/pytorch/pytorch/pull/59980.

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -35,71 +36,70 @@ using namespace at::sparse_csr;
-  s_addmm_out_csr_sparse_dense_cuda_worker(nnz, m, n, k, r, beta, t, alpha, crow_indices, col_indices, values, dense);
-  return r;
+
+  sparse::cuda::impl::addmm_out_sparse_csr_dense_cuda_impl(mat1, mat2, beta, alpha, result);
```
```suggestion
  sparse::impl::cuda::addmm_out(mat1, mat2, beta, alpha, result);
```

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.h:
```diff
@@ -0,0 +1,24 @@
+namespace at {
+namespace native {
+namespace sparse {
+namespace cuda {
+namespace impl {
```
```suggestion
namespace impl {
namespace cuda {
```

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.h:
```diff
@@ -0,0 +1,24 @@
+namespace cuda {
+namespace impl {
+
+void addmm_out_sparse_csr_dense_cuda_impl(
```
```suggestion
void addmm_out(
```

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -0,0 +1,136 @@
+
+} // anonymous namespace
+
+void addmm_out_sparse_csr_dense_cuda_impl(
```
```suggestion
void addmm_out(
```

IvanYashchuk commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -0,0 +1,136 @@
+namespace at {
+namespace native {
+namespace sparse {
+namespace cuda {
+namespace impl {
```
```suggestion
namespace impl {
namespace cuda {
```

IvanYashchuk commented on aten/src/ATen/cuda/CUDADataType.h:
```diff
@@ -0,0 +1,96 @@
+template<> inline cudaDataType getCudaDataType<float>() {
+  return CUDA_R_32F;
+}
+template<> inline cudaDataType getCudaDataType<double>() {
+  return CUDA_R_64F;
+}
+template<> inline cudaDataType getCudaDataType<c10::complex<c10::Half>>() {
+  return CUDA_C_16F;
+}
+template<> inline cudaDataType getCudaDataType<c10::complex<float>>() {
+  return CUDA_C_32F;
+}
+template<> inline cudaDataType getCudaDataType<c10::complex<double>>() {
+  return CUDA_C_64F;
+}
+
+// HIP doesn't define integral types
+#ifndef __HIP_PLATFORM_HCC__
+template<> inline cudaDataType getCudaDataType<uint8_t>() {
+  return CUDA_R_8U;
+}
+template<> inline cudaDataType getCudaDataType<int8_t>() {
+  return CUDA_R_8I;
+}
+template<> inline cudaDataType getCudaDataType<int>() {
+  return CUDA_R_32I;
+}
+#endif
+
+#if !defined(__HIP_PLATFORM_HCC__) || (defined(CUDA_VERSION) && CUDA_VERSION >= 11000)
+template<> inline cudaDataType getCudaDataType<int16_t>() {
+  return CUDA_R_16I;
+}
+template<> inline cudaDataType getCudaDataType<int64_t>() {
+  return CUDA_R_64I;
+}
+template<> inline cudaDataType getCudaDataType<at::BFloat16>() {
+  return CUDA_R_16BF;
+}
+#endif
```
From https://github.com/pytorch/pytorch/pull/59980.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

xwang233 reviewed (commented):
Thanks for the amazing work on cuSPARSE descriptors! I think this overall looks good. I've left some comments, mainly 2 points:

- The use of `operator[]` on `ArrayRef` is not always safe [(raw pointer shift impl)](https://github.com/pytorch/pytorch/blob/65f33ec85c2a7d8fb9bf582017d3170bf89e6c12/c10/util/ArrayRef.h#L217-L219). We need to check `tensor.dim()` first, or use the safer version of `tensor.size(-1)`, `tensor.stride(-1)`.
- There are several cuda version guard macro like `defined(CUDA_VERSION) && CUDA_VERSION >= 11000`, which is a bit repetitive. Can we use your macro `AT_USE_CUSPARSE_GENERIC_API` for all these code and other existing code as well?

If any functionality involves generic API on cuda 10.2 linux vs cuda 11.0 windows, we can use the generic API only on cuda >= 11.0 regardless of OS. This will make the logic cleaner. Legacy API on cuda 10.2 linux still works fine with good performance.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/cuda/CUDASparseDescriptors.h:
```diff
@@ -0,0 +1,202 @@
+
+// If a specific GPU model does not provide native support for a given data
+// type, cuSparse routines return CUSPARSE_STATUS_ARCH_MISMATCH error
+void check_supported_cuda_type(cudaDataType cuda_type) {
```
We might want to make this something more specific to sparse, this sounds like something that could easily conflict with another function in another header.

IvanYashchuk replied:
This is in an unnamed namespace, so local to this file only.

cpuhrsch replied:
Ah I see! I thought it's generally not a good idea to have anonymous namespaces in headers.

IvanYashchuk replied:
Indeed, I'll move it to the `.cpp` file.

cpuhrsch commented on aten/src/ATen/cuda/CUDASparseDescriptors.h:
```diff
@@ -0,0 +1,202 @@
+    : public CuSparseDescriptor<cusparseDnMatDescr, &cusparseDestroyDnMat> {
+ public:
+  CuSparseDnMatDescriptor(const Tensor& input) {
+    TORCH_INTERNAL_ASSERT(input.layout() == kStrided);
```
Is there value in using TORCH_CHECK here instead or are you already guarding this with more user friendly error messages at the use sites?

IvanYashchuk replied:
This class is never exposed to the Python user in any way, so all those asserts are only for developers to state the assumptions on the input tensor.

cpuhrsch commented on aten/src/ATen/cuda/CUDASparseDescriptors.h:
```diff
@@ -0,0 +1,251 @@
+ public:
+  CuSparseSpMatCsrDescriptor(const Tensor& input) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.is_sparse_csr());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
```
As above, let's double check that the user won't run into them when passing the wrong input. You could also write tests for that.

IvanYashchuk replied:
It's not possible to create a CSR matrix in PyTorch with `.dim() != 2` currently. There are tests for `torch.sparse_csr_tensor` constructor in `test_sparse_csr.py`.
The `>= 2` condition here to account for the future possible batch extension, I'll remove handling of batched CSR since that code is untested.

cpuhrsch commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -555,4 +555,22 @@ static inline bool linalg_solve_is_vector_rhs(const Tensor& input, const Tensor&
+static inline bool is_blas_compatible_column_major_order(const Tensor& input) {
+  IntArrayRef input_strides = input.strides();
+  IntArrayRef input_sizes = input.sizes();
+  auto ndim = input.dim();
```
Is it possible that they accidentally do and cause a difficult to find bug?  Would it help to add an assert here as well?

cpuhrsch commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -0,0 +1,179 @@
+#else
+  c10::MaybeOwned<Tensor> result_ = prepare_dense_matrix_for_cusparse(result);
+  c10::MaybeOwned<Tensor> mat2_ = prepare_dense_matrix_for_cusparse(mat2);
+
```
While that checks for Tensors being on a CUDA device, I don't think that checks for the actual GPU device number etc., if that matters.

cpuhrsch reviewed (changes requested):
See inline comments

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

cpuhrsch commented:
@IvanYashchuk to land this you'll need to rebase

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk commented:
> @IvanYashchuk to land this you'll need to rebase

Done. CI seems to be broken at the moment though:
 `fatal: not a git repository: /__w/pytorch/pytorch/.git/modules/third_party/protobuf`

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

cpuhrsch commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -1212,5 +1212,9 @@ def wrap_fn(self, *args, **kwargs):
     return skipCUDAIfCudnnVersionLessThan(0)(fn)
 
+# Skips a test on CUDA if cuSparse generic API is not available
+def skipCUDAIfNoCusparseGeneric(fn):
```
This seems unused

IvanYashchuk replied:
Right, moved it higher in the stack.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

IvanYashchuk commented:
Oh MSVC can't compile nested `#if #else #endif` and `win-vs2019-cuda11.3-py3 / build` fails.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk commented:
@cpuhrsch, Windows build failure was fixed.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

cpuhrsch commented on aten/src/ATen/cuda/CUDASparse.h:
```diff
@@ -0,0 +1,12 @@
+// cuSparse Generic API added in CUDA 10.1
+// Windows support added in CUDA 11.0
+// ROCm is not enabled
+#if defined(CUDART_VERSION) && defined(CUSPARSE_VERSION) && ((CUSPARSE_VERSION >= 10300) || (CUSPARSE_VERSION >= 11000 && defined(_WIN32)))
```
In the case it's not windows the CUSPARSE_VERSION here is requried to be greater than 10.3, but the [test check below](https://github.com/pytorch/pytorch/pull/60838/files#diff-fe348e24069d43bc7c6913174b038fcc5880a3281bdc0e8e217cf210bd0935e5R170-R175) tests for 10.1, if I'm not miss-understanding the conventions used by CUSPARSE_VERSION

IvanYashchuk replied:
CUDA Toolkit uses different version numbers from its components like cuBLAS, cuSOLVER, cuSPARSE. It's actually possible to have a CUDA Toolkit (compilers, etc.) be of version for example 11.4 and cuSPARSE from toolkit of 11.2 (though it'd be a rare situation)
See Table 1 here https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions
In tests, we have no way currently to distinguish between cuda toolkit versions and cusparse, it's also not possible to determine the "update/patch" version (like in `11.1.patch_version`).

cpuhrsch commented on aten/src/ATen/cuda/CUDASparseDescriptors.cpp:
```diff
@@ -0,0 +1,168 @@
+  if (ndim > 2) {
+    auto batch_count =
+        at::native::cuda_int_cast(at::native::batchCount(input), "batch_count");
+    auto batch_stride = input_strides[ndim - 3];
```
Since we can have arbitrary ndim here, why do we choose input_strides[ndim - 3] as the batch_stride?

IvanYashchuk replied:
When matrices are stored in the last two dimensions in order to iterate over all matrices in n-dimensional tensor we need to use `input_strides[ndim - 3]`.
`input_strides[ndim - 1]` is used to iterate over scalar,
`input_strides[ndim - 2]` is used to iterate over vectors,
`input_strides[ndim - 3]` is used to iterate over matrices,
```
for (int64_t batch_id = 0; batch_id < prod(tensor.shape[:-2]); batch_id++) {
    // in contiguous case tensor.stride(-3) == tensor.stride(-1)*tensor.stride(-2)
    current_matrix_ptr = tensor_data[batch_id*tensor.stride(-1)*tensor.stride(-2)]
}
```

cpuhrsch replied:
Sure, but that'd be for a 3-dim Tensor only, no? But we're checking for ndim greater than 2, so what about the 4d case?

IvanYashchuk replied:
No, it would be N-dim Tensor, where N >= 3.
The data of tensors is a 1D array in memory and elements are accessed using a single `operator[]`. The tensors can be 543d and the code will remain correct, last two dimensions are matrices and all other could as well be reshaped to a 3d tensor for iteration.
```python
In [1]: import torch

In [2]: a = torch.randn(2, 2, 2, 2) # 4d case

In [3]: a
Out[3]:
tensor([[[[-1.0814, -0.5789],
          [ 1.0179, -2.3714]],

         [[ 0.2434,  0.4329],
          [ 0.4716, -0.7596]]],


        [[[-0.4049,  1.5974],
          [-0.9816,  1.1460]],

         [[-0.1591,  1.3282],
          [-0.0039,  0.1529]]]])

In [4]: a.flatten().shape
Out[4]: torch.Size([16])

In [5]: for batch_id in range(2*2):
   ...:     batch_stride = a.stride(-3)
   ...:     current_matrix_start = a.flatten()[batch_id * batch_stride]
   ...:     print(current_matrix_start)
   ...:
tensor(-1.0814)
tensor(0.2434)
tensor(-0.4049)
tensor(-0.1591)
```

cpuhrsch replied:
Sure, but this only is true if the memory is contiguous. I guess we have a contiguity check somewhere. In any case we're only supporting 2d for now.

cpuhrsch commented on aten/src/ATen/cuda/CUDASparseDescriptors.cpp:
```diff
@@ -0,0 +1,168 @@
+      ));
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000
+  if (ndim > 2) {
```
CSR can't have dimension larger than 2 (for now), why do we have this check?

IvanYashchuk replied:
Forgot to remove it.

cpuhrsch commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -576,4 +576,24 @@ static inline bool linalg_solve_is_vector_rhs(const Tensor& input, const Tensor&
+  IntArrayRef input_strides = input.strides();
+  IntArrayRef input_sizes = input.sizes();
+  auto ndim = input.dim();
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(ndim >= 2);
```
I'm wondering in general why we often use ```ndim >= 2``` instead of ```ndim == 2```?

IvanYashchuk replied:
Was setting the stage for the batched case and forgot to change it to `ndim == 2`.

cpuhrsch replied:
Does cuSPARSE do something special for the batched case, i.e. do they have specialized kernels here or are they flattening it internally and just dispatch to the 2d-Tensor kernels after some reshaping?

IvanYashchuk replied:
It has specialized kernels for batched sparse matrix-dense matrix multiplications. From https://github.com/pytorch/pytorch/issues/60854#pytorch110-tasks:

> cuSPARSE supports batching Cᵢ = A ⋅ Bᵢ, Cᵢ = Aᵢ ⋅ B, Cᵢ = Aᵢ ⋅ Bᵢ.

Where A is a sparse matrix, B is a dense matrix, i is batch id.

The internal implementation of cuSPARSE is unknown, it's closed source.

cpuhrsch replied:
Ok we can ask NVIDIA or look at nvvp and check the trace of kernel launches on various input sizes

cpuhrsch replied:
Maybe cc @xwang233

xwang233 replied:
Batched operations are supported in cuSPARSE and may give better performance. We definitely want to support this use case in pytorch as well.

IvanYashchuk committed:
Update on "Add cuSPARSE descriptors and update CSR addmm"


Rewrote `addmm_out_sparse_csr_dense_cuda` implementation using new cusparse descriptors.

`addmm` now works without conversions with both 32-bit and 64-bit indices.
The dense tensors can have a row- or column-major layout. If the dense tensors are a contiguous slice of a larger tensor, the storage is used directly without temporary copies.

Differential Revision: [D30643191](https://our.internmc.facebook.com/intern/diff/D30643191)

cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -576,4 +576,24 @@ static inline bool linalg_solve_is_vector_rhs(const Tensor& input, const Tensor&
+  IntArrayRef input_strides = input.strides();
+  IntArrayRef input_sizes = input.sizes();
+  auto ndim = input.dim();
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(ndim == 2);
```
This might be something we'll need to revisit if https://github.com/pytorch/pytorch/issues/63583 isn't resolved. We can merge it for now, but beware.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30643191).

IvanYashchuk commented:
@ngimel, @mruberry this PR broke windows binary builds, could you please revert it? https://app.circleci.com/pipelines/github/pytorch/pytorch/387990/workflows/93142049-c52a-45e1-bbb9-ce27dcc07d0c/jobs/16286890

Why no CI jobs were launched for the merge commit https://github.com/pytorch/pytorch/commit/541eb1db639ffa159305e0fb75081eb51b5e987e ?

mruberry commented:
@IvanYashchuk It should unland soon. It looks like CI jobs aren't triggering properly at the moment. I've pinged the team.

IvanYashchuk commented:
The only broken CI job (`binary_windows_wheel_3_7_cu102_build`) was removed in https://github.com/pytorch/pytorch/pull/65961. The merge commit https://github.com/pytorch/pytorch/commit/541eb1db639ffa159305e0fb75081eb51b5e987e was not reverted, which is a good thing.
